server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java

```diff
@@ -0,0 +1,137 @@
+        String nodeWithPrimary = internalCluster().startNode();
+        assertAcked(
+            client().admin().indices().prepareCreate(indexName)
+                .setSettings(Settings.builder()
+                    .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true)
+                    .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+                    .put(IndexSettings.FILE_BASED_RECOVERY_THRESHOLD_SETTING.getKey(), 1.0f)
+                    .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "100ms")
+                    .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "1ms")));
+        String nodeWithReplica = internalCluster().startDataOnlyNode();
+        Settings nodeWithReplicaSettings = internalCluster().dataPathSettings(nodeWithReplica);
+        ensureGreen(indexName);
+        indexRandom(randomBoolean(), randomBoolean(), randomBoolean(), IntStream.range(0, between(100, 1000))
+            .mapToObj(n -> client().prepareIndex(indexName, "_doc", Integer.toString(n)).setSource("f", "v")).collect(Collectors.toList()));
+        ensureGlobalCheckpointSyncedAndPersisted(indexName);
+        if (randomBoolean()) {
+            client().admin().indices().prepareFlush(indexName).get();
+        }
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(nodeWithReplica));
+        if (randomBoolean()) {
+            indexRandom(randomBoolean(), randomBoolean(), randomBoolean(), IntStream.range(0, between(10, 200))
+                .mapToObj(n -> client().prepareIndex(indexName, "_doc", Integer.toString(n)).setSource("f", "v"))
+                .collect(Collectors.toList()));
+        }
+        if (randomBoolean()) {
+            client().admin().indices().prepareForceMerge(indexName).setFlush(true).get();
+        }
+        CountDownLatch blockRecovery = new CountDownLatch(1);
+        CountDownLatch recoveryStarted = new CountDownLatch(1);
+        MockTransportService transportServiceOnPrimary
+            = (MockTransportService) internalCluster().getInstance(TransportService.class, nodeWithPrimary);
+        transportServiceOnPrimary.addSendBehavior((connection, requestId, action, request, options) -> {
+            if (PeerRecoveryTargetService.Actions.FILES_INFO.equals(action)) {
+                recoveryStarted.countDown();
+                try {
+                    blockRecovery.await();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+        internalCluster().startDataOnlyNode();
+        recoveryStarted.await();
+        nodeWithReplica = internalCluster().startDataOnlyNode(nodeWithReplicaSettings);
+        // AllocationService only calls GatewayAllocator if there're unassigned shards
```

Comment:
I wonder if this is not important to fix as part of this PR?

If we have an ongoing recovery and we started some node that has potentially better data, we will not use it unless something external triggers it (like here). This could happen for all shards on a node? This was the same so far with the sync id, but I think it is worse now since it makes it less certain that we do ops-based recovery.

And vice-versa, there is a staleness concern too. Consider that a file based recovery has been running for a while and some node containing data has been started for a while too. If someone then triggers this by for instance creating an index, we might cancel the recovery to put it on to the node containing the data. But I think this decision will be made based on stale lease data, since we only clear the async-fetch-cache when shards are started.

Reply:
> And vice-versa, there is a staleness concern too.

@henningandersen, I opened #47035 using your work in #42518.

> Note that a fix for this might arrive via #46520

Thanks @DaveCTurner.